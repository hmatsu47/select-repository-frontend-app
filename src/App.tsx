import { Component } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import { TitleBar } from "./TitleBar";

export const App: Component = () => {
  return (
    <>
      <TitleBar />
      <Box
        sx={{
          padding: "10px 0 0 0",
          width: "100%",
          minWidth: "960px",
          display: "flex",
          justifyContent: "center",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column"></Stack>
      </Box>
    </>
  );
};
