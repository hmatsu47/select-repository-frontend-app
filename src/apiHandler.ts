import { fetchWithTimeout } from "./fetchWithTimeout";

// APIからデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout(url)).json();
};
// APIからデータ取得（POST）
export const postApiData = async (url: string, object: Object) => {
  return await (await fetchWithTimeout(url, object)).json();
};
