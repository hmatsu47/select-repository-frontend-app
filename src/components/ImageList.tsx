import { For, Show } from "solid-js";
import Box from "@suid/material/Box";
import Button from "@suid/material/Button";
import Paper from "@suid/material/Paper";
import Table from "@suid/material/Table";
import TableBody from "@suid/material/TableBody";
import TableCell from "@suid/material/TableCell";
import TableContainer from "@suid/material/TableContainer";
import TableHead from "@suid/material/TableHead";
import TableRow from "@suid/material/TableRow";
import Typography from "@suid/material/Typography";
import {
  images,
  imageUri,
  isOpenedConfirm,
  repository,
  repositoryUri,
  setImageUri,
} from "../signal";
import { formatDateTimeDisplay } from "../formatDateTime";

export const ImageList = () => {
  return (
    <Show
      when={repository() && repositoryUri() && images() !== undefined}
      fallback={<></>}
    >
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Show
          when={images() !== null}
          fallback={
            <Typography variant="h6">
              リポジトリにイメージがありません
            </Typography>
          }
        >
          <TableContainer component={Paper}>
            <Table aria-label="simple table">
              <TableHead>
                <TableRow>
                  <TableCell>選択</TableCell>
                  <TableCell>
                    イメージURI
                    <br />
                    （リポジトリURI:タグ または リポジトリURI@ダイジェスト）
                  </TableCell>
                  <TableCell>プッシュ日時</TableCell>
                  <TableCell>
                    サイズ
                    <br />
                    （MB）
                  </TableCell>
                  <TableCell>
                    脆弱性調査
                    <br />
                    （要AWSログイン）
                  </TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                <For each={images()} fallback={<></>}>
                  {(imageItem) => (
                    <TableRow>
                      <TableCell component="th" scope="row">
                        <Button
                          disabled={isOpenedConfirm()}
                          variant={
                            imageUri() && imageUri() === imageItem.uri
                              ? "contained"
                              : "outlined"
                          }
                          size="small"
                          color={
                            imageUri() && imageUri() === imageItem.uri
                              ? "primary"
                              : "inherit"
                          }
                          onClick={() => {
                            setImageUri(imageItem.uri);
                          }}
                          sx={{ textTransform: "none" }}
                          title={`${imageItem.uri} を選択`}
                        >
                          選択
                        </Button>
                      </TableCell>
                      <TableCell>{imageItem.uri}</TableCell>
                      <TableCell>
                        {formatDateTimeDisplay(new Date(imageItem.pushed_at))}
                      </TableCell>
                      <TableCell align="right">
                        {(
                          Math.round(
                            (imageItem.size / (1000 * 1000)) * Math.pow(10, 2)
                          ) / Math.pow(10, 2)
                        ).toLocaleString()}
                      </TableCell>
                      <TableCell>
                        <Button
                          disabled={isOpenedConfirm()}
                          variant="text"
                          onClick={() => {
                            const url = `https://${
                              repositoryUri()!.split(".")[3]
                            }.console.aws.amazon.com/ecr/repositories/private/${
                              repositoryUri()!.split(".")[0]
                            }/${repository()}/image/${
                              imageItem.digest
                            }/scan-results/?region=${
                              repositoryUri()!.split(".")[3]
                            }`;
                            window.open(url, "_blank");
                          }}
                          sx={{ textTransform: "none" }}
                          title={`${imageItem.digest} の結果を表示`}
                        >
                          結果表示
                        </Button>
                      </TableCell>
                    </TableRow>
                  )}
                </For>
              </TableBody>
            </Table>
          </TableContainer>
        </Show>
      </Box>
    </Show>
  );
};
