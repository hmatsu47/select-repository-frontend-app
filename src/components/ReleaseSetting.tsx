import { Show } from "solid-js";
import Box from "@suid/material/Box";
import Button from "@suid/material/Button";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { formatDateTimeInput, formatDateTimeStore } from "../formatDateTime";
import {
  images,
  imageUri,
  isReleased,
  isReleaseSelected,
  releaseAt,
  setReleaseAt,
  setIsOpenedConfirm,
  setIsCancel,
} from "../signal";

export const ReleaseSetting = () => {
  const handleOpenSet = () => {
    setIsCancel(false);
    setIsOpenedConfirm(true);
  };
  const handleOpenCancel = () => {
    setIsCancel(true);
    setIsOpenedConfirm(true);
  };
  return (
    <Show when={isReleased() !== undefined && images()} fallback={<></>}>
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column">
          <Stack
            spacing={2}
            direction="row"
            sx={{ display: "flex", alignItems: "center" }}
          >
            <Typography variant="h6" sx={{ minWidth: "256px", color: "#f00" }}>
              次回リリースイメージ URI
            </Typography>
            <Typography variant="subtitle1">
              {imageUri() ? imageUri() : "下のリストから選択してください"}
            </Typography>
          </Stack>
          <Stack spacing={2} direction="row">
            <Typography variant="h6" sx={{ minWidth: "256px", color: "#f00" }}>
              <label id="releaseAt">次回リリース日時</label>
            </Typography>
            <input
              type="datetime-local"
              aria-labelledby="releaseAt"
              value={
                releaseAt() ? formatDateTimeInput(new Date(releaseAt()!)) : ""
              }
              onChange={(event) => {
                setReleaseAt(
                  event.currentTarget.value === ""
                    ? undefined
                    : new Date(
                        Date.parse(
                          formatDateTimeStore(event.currentTarget.value)
                        )
                      )
                );
              }}
              max="9999-12-31 23:59"
            />
            <Button
              variant="contained"
              size="small"
              color="primary"
              onClick={() => {
                var tomorrow = new Date();
                tomorrow.setDate(tomorrow.getDate() + 1);
                tomorrow.setHours(4);
                tomorrow.setMinutes(5);
                tomorrow.setSeconds(0);
                tomorrow.setMilliseconds(0);
                setReleaseAt(tomorrow);
              }}
              sx={{ textTransform: "none" }}
              title="翌朝リリースする"
            >
              翌朝 04:05
            </Button>
            <Button
              variant="contained"
              size="small"
              color="secondary"
              onClick={() => {
                setReleaseAt(new Date());
              }}
              sx={{ textTransform: "none" }}
              title="即時リリースする"
            >
              現在（即時リリースする）
            </Button>
          </Stack>
          <Stack spacing={2} direction="row">
            <Typography variant="h6" sx={{ minWidth: "256px", color: "#f00" }}>
              次回リリースセット
            </Typography>
            <Button
              disabled={!imageUri() || !releaseAt()}
              variant="contained"
              size="small"
              color="primary"
              onClick={handleOpenSet}
              sx={{ textTransform: "none" }}
              title="次回リリースをセット"
            >
              指定のイメージ URI と日時をセット
            </Button>
            <Button
              disabled={!isReleaseSelected()}
              variant="contained"
              size="small"
              color="secondary"
              onClick={handleOpenCancel}
              sx={{ textTransform: "none" }}
              title="次回リリースを取り消し"
            >
              リリースを取り消し
            </Button>
          </Stack>
        </Stack>
      </Box>
    </Show>
  );
};
