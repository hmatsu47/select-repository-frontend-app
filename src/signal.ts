import { createSignal } from "solid-js";
import { AlertColor } from "@suid/material/Alert";

import { ImageItem, RepositoryItem, ServiceItem } from "./type";

export const [service, setService] = createSignal<string | undefined>();
export const [repository, setRepository] = createSignal<string | undefined>();
export const [services, setServices] = createSignal<
  ServiceItem[] | undefined
>();
export const [repositories, setRepositories] = createSignal<
  RepositoryItem[] | undefined
>();
export const [images, setImages] = createSignal<ImageItem[] | undefined>();
export const [imageUri, setImageUri] = createSignal<string | undefined>();
export const [releaseAt, setReleaseAt] = createSignal<Date | undefined>();
export const [lastImageUri, setLastImageUri] = createSignal<
  string | undefined
>();
export const [lastReleasedAt, setLastReleasedAt] = createSignal<
  Date | undefined
>();
export const [isReleased, setIsReleased] = createSignal<boolean | undefined>();
export const [message, setMessage] = createSignal<string | undefined>();
export const [messageSeverity, setMessageSeverity] = createSignal<
  AlertColor | undefined
>();
