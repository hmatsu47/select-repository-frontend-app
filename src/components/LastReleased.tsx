import { Show } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { formatDateTimeDisplay } from "../formatDateTime";
import { isReleased, lastImageUri, lastReleasedAt } from "../signal";

export const LastReleased = () => {
  return (
    <Show when={isReleased()} fallback={<></>}>
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column">
          <Stack spacing={2} direction="row">
            <Typography variant="h6" sx={{ minWidth: "256px" }}>
              前回リリースイメージURI
            </Typography>
            <Typography variant="h6">{lastImageUri()!}</Typography>
          </Stack>
          <Stack spacing={2} direction="row">
            <Typography variant="h6" sx={{ minWidth: "256px" }}>
              前回リリース日時
            </Typography>
            <Typography variant="h6">
              {formatDateTimeDisplay(new Date(lastReleasedAt()!))}
            </Typography>
          </Stack>
        </Stack>
      </Box>
    </Show>
  );
};
