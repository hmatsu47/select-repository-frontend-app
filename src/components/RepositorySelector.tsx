import { For, Show } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { RepositoryButton } from "./RepositoryButton";
import { repositories } from "../signal";

export const RepositorySelector = () => {
  return (
    <Box
      sx={{
        width: "100%",
        minWidth: "1024px",
        display: "flex",
      }}
      aria-live="polite"
    >
      <Show when={repositories()} fallback={<></>}>
        <Stack spacing={2} direction="row">
          <Typography variant="h6" sx={{ minWidth: "256px" }}>
            リリース対象リポジトリ
          </Typography>
          <For each={repositories()} fallback={<></>}>
            {(repositoryItem) => (
              <RepositoryButton repositoryItem={repositoryItem} />
            )}
          </For>
        </Stack>
      </Show>
    </Box>
  );
};
