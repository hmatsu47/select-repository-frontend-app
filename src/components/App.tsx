import { Component } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import { ImageList } from "./ImageList";
import { LastReleased } from "./LastReleased";
import { Message } from "./Message";
import { ReleaseSetting } from "./ReleaseSetting";
import { RepositorySelector } from "./RepositorySelector";
import { ServiceSelector } from "./ServiceSelector";
import { TitleBar } from "./TitleBar";

export const App: Component = () => {
  return (
    <>
      <TitleBar />
      <Box
        sx={{
          padding: "10px 10px 0 10px",
          width: "100%",
          minWidth: "1024px",
          display: "flex",
          // justifyContent: "center",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column">
          <ServiceSelector />
          <LastReleased />
          <RepositorySelector />
          <ReleaseSetting />
          <Message />
          <ImageList />
        </Stack>
      </Box>
    </>
  );
};
