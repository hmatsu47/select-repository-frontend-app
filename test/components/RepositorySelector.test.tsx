import { describe, expect, test } from "vitest";
import { render } from "solid-testing-library";
import { formatSnapshot } from "../common/formatSnapshot";
import { RepositoryItem } from "../../src/type";
import { RepositorySelector } from "../../src/components/RepositorySelector";
import { setRepositories, setService } from "../../src/signal";

describe("<RepositorySelector />", () => {
  const repositorySelectorList = [
    {
      title: "リポジトリ x1",
      service: "service1",
      repositories: [
        {
          name: "repository1",
          uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository1",
        } as RepositoryItem,
      ],
    },
    {
      title: "リポジトリ x2",
      service: "service2",
      repositories: [
        {
          name: "repository1",
          uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository1",
        } as RepositoryItem,
        {
          name: "repository2",
          uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository2",
        } as RepositoryItem,
      ],
    },
    {
      title: "リポジトリ x3",
      service: "service3",
      repositories: [
        {
          name: "repository1",
          uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository1",
        } as RepositoryItem,
        {
          name: "repository2",
          uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository2",
        } as RepositoryItem,
        {
          name: "repository3",
          uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository3",
        } as RepositoryItem,
      ],
    },
  ];
  repositorySelectorList.forEach((testCase) => {
    test(testCase.title, async () => {
      localStorage.removeItem(`selectedRepository-${testCase.service}`);
      setService(testCase.service);
      setRepositories(testCase.repositories);
      const { container, findByText, unmount } = render(() => (
        <RepositorySelector />
      ));
      // css の名前が動的に変わるので固定値に置換
      const html = formatSnapshot(container.innerHTML);
      expect(html).toMatchSnapshot();
      // 各ボタン
      testCase.repositories.forEach(async (repository) => {
        const button = (await findByText(repository.name)) as HTMLElement;
        expect(button).toHaveTextContent(repository.name);
      });
      unmount();
    });
  });
});
