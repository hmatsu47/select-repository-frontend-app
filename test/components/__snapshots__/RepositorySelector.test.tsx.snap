// Vitest Snapshot v1

exports[`<RepositorySelector /> > リポジトリ x1 1`] = `"<div aria-live=\\"polite\\" class=\\"MuiBox-root css-xxxxxx\\"><div class=\\"MuiStack-root css-xxxxxx\\"><h6 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxx\\">リリース対象リポジトリ</h6><button type=\\"button\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxx\\" title=\\"repository1 を選択\\" tabindex=\\"0\\">repository1<span class=\\"MuiTouchRipple-root MuiTouchRipple-root css-xxxxxx\\"></span></button></div></div>"`;

exports[`<RepositorySelector /> > リポジトリ x2 1`] = `"<div aria-live=\\"polite\\" class=\\"MuiBox-root css-xxxxxx\\"><div class=\\"MuiStack-root css-xxxxxx\\"><h6 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxx\\">リリース対象リポジトリ</h6><button type=\\"button\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxx\\" title=\\"repository1 を選択\\" tabindex=\\"0\\">repository1<span class=\\"MuiTouchRipple-root MuiTouchRipple-root css-xxxxxx\\"></span></button><button type=\\"button\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxx\\" title=\\"repository2 を選択\\" tabindex=\\"0\\">repository2<span class=\\"MuiTouchRipple-root MuiTouchRipple-root css-xxxxxx\\"></span></button></div></div>"`;

exports[`<RepositorySelector /> > リポジトリ x3 1`] = `"<div aria-live=\\"polite\\" class=\\"MuiBox-root css-xxxxxx\\"><div class=\\"MuiStack-root css-xxxxxx\\"><h6 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxx\\">リリース対象リポジトリ</h6><button type=\\"button\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxx\\" title=\\"repository1 を選択\\" tabindex=\\"0\\">repository1<span class=\\"MuiTouchRipple-root MuiTouchRipple-root css-xxxxxx\\"></span></button><button type=\\"button\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxx\\" title=\\"repository2 を選択\\" tabindex=\\"0\\">repository2<span class=\\"MuiTouchRipple-root MuiTouchRipple-root css-xxxxxx\\"></span></button><button type=\\"button\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxx\\" title=\\"repository3 を選択\\" tabindex=\\"0\\">repository3<span class=\\"MuiTouchRipple-root MuiTouchRipple-root css-xxxxxx\\"></span></button></div></div>"`;
